src/mame/drivers/enmirage.cpp

```diff
@@ -103,144 +151,288 @@ void enmirage_state::floppy_formats(format_registration &fr)
+    PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("7")        PORT_CODE(KEYCODE_7)    PORT_CHANGED_MEMBER(DEVICE_SELF, enmirage_state, input_changed, 0)
+    PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("On/Up")    PORT_CODE(KEYCODE_UP)   PORT_CHANGED_MEMBER(DEVICE_SELF, enmirage_state, input_changed, 0)
+    PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Param")    PORT_CODE(KEYCODE_I)    PORT_CHANGED_MEMBER(DEVICE_SELF, enmirage_state, input_changed, 0)
+    PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Off/Down") PORT_CODE(KEYCODE_DOWN) PORT_CHANGED_MEMBER(DEVICE_SELF, enmirage_state, input_changed, 0)
+    PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Value")    PORT_CODE(KEYCODE_J)    PORT_CHANGED_MEMBER(DEVICE_SELF, enmirage_state, input_changed, 0)
+    PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Cancel")   PORT_CODE(KEYCODE_K)    PORT_CHANGED_MEMBER(DEVICE_SELF, enmirage_state, input_changed, 0)
+
+    PORT_START(PITCH_TAG)
+    PORT_BIT( 0xff, 0x7f, IPT_PADDLE) PORT_NAME("Pitch Wheel") PORT_SENSITIVITY(100) PORT_KEYDELTA(10) PORT_MINMAX(0x00,0xff) PORT_CODE_INC(KEYCODE_4_PAD) PORT_CODE_DEC(KEYCODE_1_PAD) PORT_PLAYER(1)
+    PORT_START(MOD_TAG)
+    PORT_BIT( 0xff, 0x7f, IPT_PADDLE) PORT_NAME("Mod Wheel") PORT_SENSITIVITY(100) PORT_KEYDELTA(10) PORT_MINMAX(0x00,0xff) PORT_CODE_INC(KEYCODE_6_PAD) PORT_CODE_DEC(KEYCODE_3_PAD) PORT_PLAYER(1)
 INPUT_PORTS_END
 
 ROM_START( enmirage )
-	ROM_REGION(0x1000, "osrom", 0)
-	ROM_LOAD( "mirage.bin",   0x0000, 0x1000, CRC(9fc7553c) SHA1(ec6ea5613eeafd21d8f3a7431a35a6ff16eed56d) )
+    ROM_REGION(0x1000, "osrom", 0)
+    ROM_LOAD( "mirage.bin", 0x0000, 0x1000, CRC(9fc7553c) SHA1(ec6ea5613eeafd21d8f3a7431a35a6ff16eed56d) )
 
-	ROM_REGION(0x20000, "es5503", ROMREGION_ERASE)
+    ROM_REGION(0x20000, "es5503", ROMREGION_ERASE)
 ROM_END
 
 void enmirage_state::init_mirage()
 {
-	floppy_image_device *floppy = m_floppy_connector ? m_floppy_connector->get_device() : nullptr;
-	if (floppy)
-	{
-		m_fdc->set_floppy(floppy);
-
-		floppy->ss_w(0);
-	}
-
-	// port A: front panel
-	m_via->write_pa0(0);
-	m_via->write_pa1(0);
-	m_via->write_pa2(0);
-	m_via->write_pa3(0);
-	m_via->write_pa4(0);
-	m_via->write_pa5(0);
-	m_via->write_pa6(0);
-	m_via->write_pa7(0);
-
-	// port B:
-	//  bit 6: IN FDC disk ready
-	//  bit 5: IN 5503 sync (?)
-	m_via->write_pb0(0);
-	m_via->write_pb1(0);
-	m_via->write_pb2(0);
-	m_via->write_pb3(0);
-	m_via->write_pb4(0);
-	m_via->write_pb5(1);
-	m_via->write_pb6(0);    // how to determine if a disk is inserted?
-	m_via->write_pb7(0);
+    floppy_image_device *floppy = m_floppy_connector ? m_floppy_connector->get_device() : nullptr;
+    if (floppy)
+    {
+        m_fdc->set_floppy(floppy);
+    }
 }
 
-CONS( 1984, enmirage, 0, 0, mirage, mirage, enmirage_state, init_mirage, "Ensoniq", "Mirage", MACHINE_NOT_WORKING )
+CONS( 1984, enmirage, 0, 0, mirage, mirage, enmirage_state, init_mirage, "Ensoniq", "Mirage", MACHINE_IS_INCOMPLETE )
```

Comment:
`MACHINE_IS_INCOMPLETE` is for machines that are properly emulated but represent incomplete prototypes.  Please use flags properly.